File: web/src/lib/api/roster.ts
import { api } from './client'

export type PaymentSituation = 'pago' | 'parcial' | 'a_pagar' | 'atrasado' | 'sem_cobranca'

export const SITUATION_LABEL: Record<PaymentSituation, string> = {
  pago: 'Pago',
  parcial: 'Parcial',
  a_pagar: 'A pagar',
  atrasado: 'Atrasado',
  sem_cobranca: 'Sem cobrança',
}

export const SITUATION_COLOR: Record<PaymentSituation, string> = {
  pago: 'success',
  parcial: 'warning',
  a_pagar: 'medium',
  atrasado: 'danger',
  sem_cobranca: 'light',
}

export interface RequirementItem {
  id: number
  requirement_id: number
  name: string
  required: boolean
  delivered: boolean
}

export interface RosterRow {
  id: number
  name: string
  email: string
  phone: string
  status: string
  is_minor: boolean
  guardian_name: string
  guardian_phone: string
  shirt_size: string
  boarding_point: string
  room_group: string
  dietary_restrictions: string
  medical_notes: string
  payment: {
    total: string
    paid: string
    remaining: string
    installments_count: number
    installments_paid: number
    overdue_count: number
    next_due_date: string | null
    situation: PaymentSituation
  }
  requirements: {
    total: number
    delivered: number
    required_total: number
    required_pending: string[]
    all_required_delivered: boolean
    items: RequirementItem[]
  }
}

export interface RosterSummary {
  participants: number
  capacity: number | null
  spots_left: number | null
  paid: number
  partial: number
  to_pay: number
  overdue: number
  minors: number
  pending_requirements: number
  total_expected: string
  total_received: string
  total_remaining: string
}

export interface Roster {
  summary: RosterSummary
  participants: RosterRow[]
}

export const getRoster = (tripId: number) => api.get<Roster>(`/trips/${tripId}/roster/`)

export interface Installment {
  id: number
  payment: number
  participant_name: string
  amount: string
  due_date: string
  status: 'pending' | 'paid' | 'overdue' | 'failed'
  pix_qr_code: string
  paid_at: string | null
}

export interface Payment {
  id: number
  participant: number
  participant_name: string
  trip: number
  total_amount: string
  status: string
  installments: Installment[]
}

interface Paginated<T> {
  count: number
  results: T[]
}

export const listPayments = (tripId: number) =>
  api.get<Paginated<Payment>>(`/payments/?trip=${tripId}`).then((p) => p.results)

export const payInstallment = (id: number) => api.post<Installment>(`/installments/${id}/pay/`)
export const unpayInstallment = (id: number) => api.post<Installment>(`/installments/${id}/unpay/`)
export const generatePix = (id: number) => api.post<Installment>(`/installments/${id}/pix/`)

/** Re-parcela o saldo em aberto (mais ou menos parcelas). */
export const replanPayment = (paymentId: number, installments: number) =>
  api.post<Payment>(`/payments/${paymentId}/replan/`, { installments })

export const setRequirementDelivered = (statusId: number, delivered: boolean) =>
  api.patch<RequirementItem>(`/participant-requirements/${statusId}/`, { delivered })

// --- inscrição pública -------------------------------------------------------

export interface EnrollmentInput {
  name: string
  email?: string
  phone?: string
  document?: string
  birth_date?: string | null
  is_minor: boolean
  guardian_name?: string
  guardian_phone?: string
  emergency_contact?: string
  health_insurance?: string
  dietary_restrictions?: string
  medical_notes?: string
  shirt_size?: string
  boarding_point?: string
  notes?: string
  consent_accepted: boolean
  installments: number
}

export interface EnrollmentResult {
  id: number
  name: string
  trip: string
  total_amount: string
  installments: { amount: string; due_date: string }[]
  requirements: string[]
  spots_left: number | null
}

export const enroll = (slug: string, input: EnrollmentInput) =>
  api.post<EnrollmentResult>(`/public/trips/${slug}/enroll/`, input)

File: api/apps/participants/services/roster.py
"""Painel de gestão da viagem — a "planilha", porém consultável e agregada.

Reúne, por participante: dados de contato, situação de pagamento (pago, a pagar,
parcelas, atraso) e a entrega dos requisitos (autorização, documentos...).
"""

from __future__ import annotations

import datetime as dt
from decimal import Decimal

from apps.payments.models import Installment
from apps.trips.models import Trip

from ..models import Participant

ZERO = Decimal("0.00")


def _payment_view(participant: Participant, hoje: dt.date) -> dict:
    """Situação financeira consolidada de um participante."""
    pagamentos = list(participant.payments.all())
    parcelas: list[Installment] = [p for pg in pagamentos for p in pg.installments.all()]

    total = sum((pg.total_amount for pg in pagamentos), ZERO)
    pago = sum((p.amount for p in parcelas if p.status == Installment.Status.PAID), ZERO)
    restante = max(total - pago, ZERO)

    em_aberto = [p for p in parcelas if p.status != Installment.Status.PAID]
    vencidas = [p for p in em_aberto if p.due_date < hoje]
    proxima = min((p.due_date for p in em_aberto), default=None)

    if total == ZERO:
        situacao = "sem_cobranca"
    elif pago >= total:
        situacao = "pago"
    elif vencidas:
        situacao = "atrasado"
    elif pago > ZERO:
        situacao = "parcial"
    else:
        situacao = "a_pagar"

    return {
        "total": total,
        "paid": pago,
        "remaining": restante,
        "installments_count": len(parcelas),
        "installments_paid": len(parcelas) - len(em_aberto),
        "overdue_count": len(vencidas),
        "next_due_date": proxima,
        "situation": situacao,
    }


def _requirements_view(participant: Participant) -> dict:
    status = list(participant.requirement_status.all())
    obrigatorios = [s for s in status if s.requirement.required]
    pendentes = [s.requirement.name for s in obrigatorios if not s.delivered]
    return {
        "total": len(status),
        "delivered": sum(1 for s in status if s.delivered),
        "required_total": len(obrigatorios),
        "required_pending": pendentes,
        "all_required_delivered": not pendentes,
        "items": [
            {
                "id": s.id,
                "requirement_id": s.requirement_id,
                "name": s.requirement.name,
                "required": s.requirement.required,
                "delivered": s.delivered,
            }
            for s in status
        ],
    }


def build_roster(trip: Trip) -> list[dict]:
    """Uma linha por participante, com pagamento e requisitos consolidados."""
    hoje = dt.date.today()
    participantes = (
        trip.participants.prefetch_related(
            "payments__installments", "requirement_status__requirement"
        )
        .all()
    )
    linhas = []
    for p in participantes:
        pagamento = _payment_view(p, hoje)
        requisitos = _requirements_view(p)
        linhas.append(
            {
                "id": p.id,
                "name": p.name,
                "email": p.email,
                "phone": p.phone,
                "status": p.status,
                "is_minor": p.is_minor,
                "guardian_name": p.guardian_name,
                "guardian_phone": p.guardian_phone,
                "shirt_size": p.shirt_size,
                "boarding_point": p.boarding_point,
                "room_group": p.room_group,
                "dietary_restrictions": p.dietary_restrictions,
                "medical_notes": p.medical_notes,
                "payment": pagamento,
                "requirements": requisitos,
            }
        )
    return linhas


def build_summary(roster: list[dict], trip: Trip) -> dict:
    """Números do topo do painel: quem pagou, quem deve, o que falta entregar."""
    def conta(situacao: str) -> int:
        return sum(1 for linha in roster if linha["payment"]["situation"] == situacao)

    total_esperado = sum((linha["payment"]["total"] for linha in roster), ZERO)
    total_recebido = sum((linha["payment"]["paid"] for linha in roster), ZERO)

    return {
        "participants": len(roster),
        "capacity": trip.capacity,
        "spots_left": (
            None
            if trip.capacity is None
            else max(
                trip.capacity
                - sum(1 for linha in roster if linha["status"] != "cancelled"),
                0,
            )
        ),
        "paid": conta("pago"),
        "partial": conta("parcial"),
        "to_pay": conta("a_pagar"),
        "overdue": conta("atrasado"),
        "minors": sum(1 for linha in roster if linha["is_minor"]),
        "pending_requirements": sum(
            1 for linha in roster if not linha["requirements"]["all_required_delivered"]
        ),
        "total_expected": total_esperado,
        "total_received": total_recebido,
        "total_remaining": max(total_esperado - total_recebido, ZERO),
    }
